### PR TITLE
Adjust handler code to prevent orphan uploads from being created [delivers #156282874]

### PR DESCRIPTION
**Pull request by jim**
## Description

The previous implementation created a row in the `uploads` table before attempting to store a file on S3, which meant that any errors raised during the upload would leave an invalid row in the database.

We had a few ideas about how to proceed here, including using database transactions, but it turns out that specifying an ID on a Pop model before it created is possible. So we now generate an ID for the new upload, use that to build a key and push the file to S3, and as a last step create the `Upload` in the database. This way, errors that are encountered can simply be returned without having to deal with cleanup (and we avoid needing a database transaction).

## Reviewer Notes

- None of the validation on `Upload` depend on the state of the database, so performing validation before the push to S3 and then skipping it later when it comes time to create the `Upload` isn't an issue.

## Verification Steps

* [ ] All tests pass.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/156282874) for this change

### CODE REVIEW

**Comment by jim** [comment]
@ntwyman @akostibas cc'ing you as this came up in planning. The solution ended up not requiring a transaction after all, so I don't think we need to discuss alternatives.

File: pkg/handlers/uploads.go [hunk 1]
```diff
@@ -68,30 +70,37 @@ func (h CreateUploadHandler) Handle(params uploadop.CreateUploadParams) middlewa
...
+		return uploadop.NewCreateUploadInternalServerError()
+	}
+
+	// Already validated upload, so just save
+	err = h.db.Create(&newUpload)
+	if err != nil {
+		h.logger.Error("DB Insertion", zap.Error(err))
+		return uploadop.NewCreateUploadInternalServerError()
+	}
+
```
Comment (kilbergr) [inline pkg/handlers/uploads.go]: Can you add back a variation of this log once it's been uploaded and saved?: `h.logger.Infof("created an upload with id %s, s3 id %s\n", newUpload.ID, newUpload.ID)`

File: pkg/models/upload_test.go [hunk 2]
```diff
@@ -39,6 +41,46 @@ func (suite *ModelSuite) Test_UploadCreate() {
 	}
 }
```
Comment (kilbergr) [inline pkg/models/upload_test.go]: Could you write a test intended to fail--like make it impossible to save to db and watch it fail the process?

File: pkg/handlers/uploads_test.go [hunk 3]
```diff
@@ -116,3 +135,24 @@ func (suite *HandlerSuite) TestCreateUploadsHandler() {
 
 	// TODO verify Body
 }
+
+func (suite *HandlerSuite) TestCreateUploadsHandlerFailuer() {
```
Comment (kilbergr) [inline pkg/handlers/uploads_test.go]: Great! Typo in test name Failuer should be Failure.

File: pkg/handlers/uploads_test.go [hunk 4]
```diff
@@ -135,3 +135,24 @@ func (suite *HandlerSuite) TestCreateUploadsHandler() {
...
+	}
+
+	count, err := suite.db.Count(&models.Upload{})
+
+	if err != nil {
+		t.Fatalf("Couldn't count uploads in database: %s", err)
+	}
+
+	if count != 0 {
+		t.Fatalf("Wroung number of uploads in database: expected 0, got %d", count)
```
Comment (kilbergr) [inline pkg/handlers/uploads_test.go]: Typo `wrong` not `wroung`

**Pull request merged**